### PR TITLE
Allow type inclusion of non-abstract objects

### LINKED ISSUE #23359: The object type can include non-abstract object type
Refer Commit https://github.com/ballerina-platform/ballerina-spec/commit/a60d2cb419000b629d9ce993a241e1c43b4d203d

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose
Fixes #23359

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -1431,7 +1432,7 @@ private void defineMembers(List<BLangTypeDefinition> typeDefNodes, SymbolEnv pkg
                     defineNode(f, objMethodsEnv);
                 });
 
-                List<String> referencedFunctions = new ArrayList<>();
+                Set<String> referencedFunctions = new HashSet<>();
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1435]: When you replace, ArrayList with a HashSet, you lose the insertion order. That may cause intermittent test failures. Better option is to replace it with a LinkedHashSet.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/object/ObjectInBaloTest.java [hunk 2]
```diff
@@ -571,7 +571,9 @@ public void testObjectReferingTypeFromBaloNegative() {
         int i = 0;
         BAssertUtil.validateError(result, i++, "undefined field 'name' in object 'Manager1'", 25, 13);
         BAssertUtil.validateError(result, i++, "undefined field 'age' in object 'Manager1'", 26, 13);
-        BAssertUtil.validateError(result, i, "incompatible types: 'foo:Manager1' is not an abstract object", 38, 6);
+        BAssertUtil.validateError(result, i,
+                                  "incompatible type reference 'foo:Manager1': a referenced object cannot have " +
+                                          "non-public fields or methods", 38, 6);
     }
```
Comment (hasithaa) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/object/ObjectInBaloTest.java:578]: We need to add a balo test case for this feature.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 1 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1435]: This is only used to check duplicates and issue ```DiagnosticCode.REDECLARED_SYMBOL``` error message.
https://github.com/ballerina-platform/ballerina-lang/blob/ac9058f25bd2282a6d4a93d4905743bc026aef7c/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java#L1879

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-type-reference.bal [hunk 3]
```diff
@@ -13,6 +13,7 @@
 // KIND, either express or implied.  See the License for the
 // specific language governing permissions and limitations
 // under the License.
+import ballerina/test;
```
[the remark this answers is not in the document]
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-type-reference.bal:16]: Will remove this.

[the remark this answers is not in the document]
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-type-reference.bal]: Done : f5f5a222febd5fa5e52b1786058666918333aaf9

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/object/ObjectInBaloTest.java (hunk 2 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/object/ObjectInBaloTest.java:594]: Added : 8e8601896a6850ade29a643e2e92a1153f220a2e

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 1 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: I renamed the DS to make sure it is not mistaken with `referencedFunctions` member field in `BObjectTypeSymbol`. 
Commit : 6e55b9358886a9cc9ec6bef3fbd9a41a834bb1d2

**Pull request merged**